### ISSUE TITLE
请问对UDP的支持怎么样？

### ISSUE THREAD

**Issue by 556a** [haxpor/Potatso#5]
如题

**Comment by haxpor** [comment]
Hey thanks for asking!
I don't see any other shadowsock client did support UDP before? Or did I miss anything?
I think there might be a reason that they won't do it.

**Comment by 556a** [comment]
Thank you

**Comment by icodesign** [comment]
We should find a way to support it but I didn't do it because of time. @haxpor

**Comment by haxpor** [comment]
Thanks @icodesign ! This confirms me that it's technical possible to do so. I noted it.

**Comment by jiangyancong** [comment]
How about enable udp relay in shadowsocks client and server? Does it work?  @haxpor

**Comment by happyfire** [comment]
I code udp relay myself in the outside of tun2socks, the ss local server send me the right udp port, but when I send data to it, I can not get any response data... I debug in ss local and can't find the udp callback called.

**Comment by jiangyancong** [comment]
@happyfire Have you enable udp-relay in shadowsocks profile? Potatso doesn't enable udp-relay by default. But I am not sure whether it works if I set the `mode` to 1.

```
typedef struct {
    /*  Required  */
   .....
    /*  Optional, set NULL if not valid   */
    char *acl;            // file path to acl
    char *log;            // file path to log
    int use_sys_log;            // file path to log
    int fast_open;        // enable tcp fast open
    int mode;             // enable udp relay
    int auth;             // enable one-time authentication
    int mtu;              // MTU of interface
    int mptcp;            // enable multipath TCP
    int verbose;          // verbose mode
} profile_t;`

**Comment by happyfire** [comment]
@jiangyancong Yes, I enabled the UDP relay in ss local. And I test the udp relay with a ss local server located in a server in LAN, the udp relay is OK. so I think the ss local used in ss-libev in potatso should have some problem.

**Comment by jiangyancong** [comment]
@happyfire I have checked the code base of ss-libev, it is almost the same with the one potatso uses. May I ask how do you pass the udp socket fd to local.c? Is it the same udp socket created in TunnelInterface.m?

**Comment by happyfire** [comment]
@jiangyancong 
The udp socket created for udp relay in ss is in udprelay.c, you can search
 `// Bind to port
    int serverfd = create_server_socket(server_host, server_port);`
I saved the serverfd here in a extern global var, so in local.c, I can use it. And remember, it is the server side udp socket fd (ss local server).
You said the udp socket in TunnelInterface.m, it is `GCDAsyncUdpSocket *udpSocket;` ?
It is for udp relay for system stack, not for ss udp relay. (and of course it is client side)
 
Actually, potatso doesn't use udp relay of ss, it just send the udp packets along system network stack, which is do in handleUDPPacket and use the udpSocket.
To relay udp using ss, you should create your own udpSocket and make sock5 select method and udp relay request tcp by yourself, and that you can send udp relay packs to ss udp socket, which is the serverfd above.

**Comment by jiangyancong** [comment]
@happyfire Thanks!  It help me a lot.

**Comment by happyfire** [comment]
@jiangyancong 
You are welcome!
I missed one thing to say, I removed antinat, and it seems that antinat doesn't support udp chain. I see from the code `We don't know how to chain UDP yet.` So If you still have problem with UDP, perhaps you should try remove antinat.

**Comment by jiangyancong** [comment]
@happyfire Could you explain more about antinat server? How to remove antinat server to support UDP request? I found the antinat server has coupling with privoxy. But someone said it's totally unnecessary to integrate antinat socket server in Potatso. what do you think?

**Comment by happyfire** [comment]
@jiangyancong 
Yes, antinat is unnecessary, for the purpose of shadowsocks proxy, there is no need to use antinat, it just do socks proxy chaining in potatso. What you found is also right, privoxy has coupling with antinat in potatso, I haven't researched it deeply, but If you want to remove antinat, it is easy.
But antinat is not used with udp, so you can still use it if you want to do udp relay.
Let's see the data flow:
For http/s:
TUN (http) -> privoxy ->ss
For tcp except http/s:
TUN (ip) ->tun2socks->antinat->ss (antinat do  socks proxy chaining, so can be removed)
For udp:
TUN (ip) -> system stack (udp) (use udpSocket)
If you want to support udp relay:
TUN(ip) -> ss (you can use another udpSocket but before that you should use a tcp to say hello to ss)

**Comment by so898** [comment]
@happyfire 
I have just get the idea of the Antinat.
The Potatso used Polipo for HTTP proxy, because the Polipo do not have function to deal with HTTP rules, so @icodesign use Antinat for rules function.
After the author use Privoxy for HTTP proxy instead of Polipo, the Antinat is not necessary anymore.

**Comment by icodesign** [comment]
@so898 You're right. Aninat is not mandatory if you handle direct tcp connections by yourself.

**Comment by Sky891028** [comment]
@happyfire 
I have try to say hello to ss_local for udp relay 
The tcp connect will close.

> 
![1502272223909](https://user-images.githubusercontent.com/6927258/29115740-3dfd7d60-7d2b-11e7-899d-73e593255a20.jpg)



need help!

**Comment by aishuidedabai** [comment]
我发现tcp和udp分开处理，tcp是通过provixy来匹配规则，udp未作处理，用114.114.114.114，53端口。所以udp是走不通的，tcp通过规则发送到tunl2socks，但是我发现在安卓上是通过dns解析来分流pdnsd，我想知道potasto2是否用的pdnsd来处理 @icodesign

**Comment by Sky891028** [comment]
how to count traffic from ssLocal ？

**Comment by so898** [comment]
@Sky891028 
I have solved this problem. It includes modifying shadowsocks-libv and ProxyManager in ProviderTunnel.
The basic idea is port the traffic record function which shadowsocks-libv used in Android version to the ProviderTunnel, then transfer the date from Tunnel Process to App Process via MMWormhole.
I might make a pull request later, but I think I will complete the new protocol support and swift 4 upgrade first.

**Comment by Sky891028** [comment]
That must be very great . @so898 
I'm really looking forward to it

**Comment by aishuidedabai** [comment]
计算流量只需要计算受代理部分的ip包，拆包计算所占字节，由于tcp是流写入和写出，所以只需要在write和read部分相加即可。udp问题，potasto1的大udp包（9216）其实是ios限制，要改掉gcdudpSocketd的最大支持栈，最新协议更简单了，更换整个live包即可，虽然我都实现了，但是里面很多很复杂的部分我还没弄懂

**Comment by Sky891028** [comment]
@aishuidedabai  
http/s的流量在哪里拿呢？ tcp&udp可以处理

**Comment by aishuidedabai** [comment]
TunnelInterface中 handleUDPPacket  handleTCPPPacket，作者进行了拆包，不用看http，因为除了http还有其他协议的流量。所以你应该算受到代理的tcp流量。在Privoxy这个服务器内你需要排除掉走国内的ip(如果你开启黑名单模式)。udp就不要算了，都走的本地

**Comment by Sky891028** [comment]
@aishuidedabai  
我应该在哪里埋点 -。-

**Comment by aishuidedabai** [comment]
我想知道Antinat这个类是不是多余的，为什么不像安卓那样直接转发给ss local，本身ss就是socks5服务器啊。。。我认为这是多余的，另外我重构整个potaso核心部分，只拿出需要的部分，重写逻辑，因为我感觉目前的逻辑复杂了。。。监听的端口应该不需要这么麻烦，如果我成功了，我会重新发布一个项目。。。非常感谢icodesign。另外swift部分都是UI，可以不用看

**Comment by aishuidedabai** [comment]
我是在live包发送sslocal这里统计所有代理流量，发出和收回

**Comment by Sky891028** [comment]
local.c这个文件吗？
如果是 我在这些方法内都没有监听到数据
static void server_recv_cb(EV_P_ ev_io *w, int revents);
static void server_send_cb(EV_P_ ev_io *w, int revents);
static void remote_recv_cb(EV_P_ ev_io *w, int revents);
static void remote_send_cb(EV_P_ ev_io *w, int revents);
如果不是 麻烦指出位置

**Comment by aishuidedabai** [comment]
remote->buf->len -= s; 是buf长度减去已发送长度    remote->buf->idx  = s; 将当前索引移动到s

**Comment by Sky891028** [comment]
呃 断点进去了  
println没有输出到控制台。。  你们是怎么看c的打印的 。。

**Comment by aishuidedabai** [comment]
没数据？不可能把，你试试直接po

**Comment by Sky891028** [comment]
po是oK的 我在代码中加入printfln(x) 系统控制台- 非xcode控制台 没有输出这些内容
